File: 6. semestr/[IBT] Bakalářská práce/Implementace/frontend/src/app/services/originalrecord.service.ts
/*******************************************************************/
/*  Project: Bachelor's thesis – Genealogy Database System         */
/*           Faculty of Information Technology                     */
/*           Brno University of Technology                         */
/*                                                                 */
/*  File:    originalrecord.service.ts                             */
/*  Date:    2019 – 2020                                           */
/*  Author:  Martin Chládek <xchlad16@stud.fit.vutbr.cz>           */
/*******************************************************************/

import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { BaseService } from './base.service';
import { QueryService } from './http/query.service';
import { OriginalRecord } from '../models/originalRecord';


@Injectable({
  providedIn: 'root'
})
export class OriginalRecordService extends BaseService{

  originalrecordUrl = "originalrecord";

  constructor(protected queryService: QueryService) { super(queryService);}
  
  getOriginalRecordsByPersonId(id: number): Observable<OriginalRecord[]> {
    const url = `${this.originalrecordUrl}/${id}`;
    return this.queryService.get<OriginalRecord[]>(url).pipe(map(result => result['recordsGet']));
  }
}

File: 6. semestr/[IBT] Bakalářská práce/Implementace/backend/Services/OriginalRecordService .cs
/*******************************************************************/
/*  Project: Bachelor's thesis – Genealogy Database System         */
/*           Faculty of Information Technology                     */
/*           Brno University of Technology                         */
/*                                                                 */
/*  File:    OriginalRecordService.cs                              */
/*  Date:    2019 – 2020                                           */
/*  Author:  Martin Chládek <xchlad16@stud.fit.vutbr.cz>           */
/*******************************************************************/

using Entity;
using Microsoft.EntityFrameworkCore;
using Services.Base;
using Services.Exceptions;
using Services.Interfaces;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Services
{
    public class OriginalRecordService : BaseService<OriginalRecord>, IOriginalRecordService
    {
        public OriginalRecordService(IServiceDependencies _serviceDependencies) : base(_serviceDependencies)
        {
        }

        #region CRUD operations

        public async Task<IEnumerable<OriginalRecord>> GetOriginalRecordsByPersonIdAsync(int id)
        {
            if (!(ExistsInDb<Person>(id)))
                throw new ObjectNotFoundException(nameof(Person), id);

            var records = await _serviceDependencies._context.OriginalRecords.Where(record => record.PersonId == id).ToListAsync();

            return records;
        }

        public async Task<OriginalRecord> CreateOriginalRecordAsync(OriginalRecord recordToCreate)
        {
            if (!(ExistsInDb<Person>(recordToCreate.PersonId)))
                throw new ObjectNotFoundException(nameof(Person), recordToCreate.PersonId);

            return await Create(recordToCreate);
        }

        public async Task DeleteAllOriginalRecordsForPersonAsync(int personId)
        {
            var records = await GetOriginalRecordsByPersonIdAsync(personId);

            foreach(var record in records)
                _serviceDependencies._context.OriginalRecords.Remove(record);

            return;
        }

        #endregion
    }
}
